File: preftils/src/main/java/com/hubble/preftils/PreferenceUtils.java
package com.hubble.preftils;

import android.content.SharedPreferences;

import org.jetbrains.annotations.NotNull;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

public class PreferenceUtils {
    static class Preference<T extends Serializable> {
        private final String key;
        private final T defaultValue;

        /**
         * Wrapper class for handling SharedPreferences
         *
         * @param key the key for the SharedPreference
         * @param defaultValue the default value for the SharedPreference
         */
        Preference(String key, @NotNull T defaultValue) {
            this.key = key;
            this.defaultValue = defaultValue;
        }

        public String getKey() {
            return key;
        }

        public T getDefault() {
            return defaultValue;
        }

        /**
         * Instance helper that wraps PreferenceUtils.get()
         * @see PreferenceUtils#get(SharedPreferences, Preference) 
         */
        public T getValue(SharedPreferences preferences) throws IOException, ClassNotFoundException {
            return get(preferences, this);
        }

        /**
         * Instance helper that wraps PreferenceUtils.put()
         * @see PreferenceUtils#put(SharedPreferences.Editor, Preference, Serializable) 
         */
        public void putValue(SharedPreferences.Editor editor, @NotNull T value) throws IOException {
            put(editor, this, value);
        }
    }

    /**
     * Retrieves the value of a preference in a type-safe way
     * <p>
     * Propagates errors thrown by ObjectInputStream.readObject() if decoding is impossible
     *
     * @param preferences instance of SharedPreferences to read from
     * @param preference the preference to read
     * @return the value retrieved in SharedPreferences, or the default value of the [preference] parameter
     *
     * @throws  ClassNotFoundException Class of a serialized object cannot be
     *          found.
     * @throws  IOException Any of the usual Input/Output related exceptions.
     *
     * @see     ObjectInputStream#readObject()
     */
    @SuppressWarnings("unchecked")
    static <T extends Serializable> T get(SharedPreferences preferences, Preference<T> preference) throws IOException, ClassNotFoundException {
        if (preference.getDefault() instanceof Integer) {
            return (T) (Integer) preferences.getInt(preference.getKey(), (Integer) preference.getDefault());
        } else if (preference.getDefault() instanceof Long) {
            return (T) (Long) preferences.getLong(preference.getKey(), (Long) preference.getDefault());
        } else if (preference.getDefault() instanceof Float) {
            return (T) (Float) preferences.getFloat(preference.getKey(), (Float) preference.getDefault());
        } else if (preference.getDefault() instanceof Boolean) {
            return (T) (Boolean) preferences.getBoolean(preference.getKey(), (Boolean) preference.getDefault());
        } else if (preference.getDefault() instanceof String) {
            return (T) preferences.getString(preference.getKey(), (String) preference.getDefault());
        } else {
            String string = preferences.getString(preference.getKey(), null);

            if (string == null) {
                return preference.getDefault();
            }

            ByteArrayInputStream inStream = new ByteArrayInputStream(string.getBytes());
            ObjectInputStream in = new ObjectInputStream(inStream);

            T object = (T) in.readObject();

            inStream.close();
            in.close();

            return object;
        }
    }

    /**
     * Sets the value of a preference in a type-safe way
     * <p>
     * Propagates errors thrown by ObjectOutputStream.writeObject() if encoding is impossible
     *
     * @param editor Instance of SharedPreferences.Editor to write to
     * @param preference the Preference to set
     * @param value the value to set the Preference to
     *
     * @throws  IOException Any exception thrown by the underlying
     *          OutputStream.
     *
     * @see     ObjectOutputStream#writeObject(Object) ()
     */
    static <T extends Serializable> void put(SharedPreferences.Editor editor, Preference<T> preference, @NotNull T value) throws IOException {
        if (value instanceof Integer) {
            editor.putInt(preference.getKey(), (Integer) value);
        } else if (value instanceof Long) {
            editor.putLong(preference.getKey(), (Long) value);
        } else if (value instanceof Float) {
            editor.putFloat(preference.getKey(), (Float) value);
        } else if (value instanceof Boolean) {
            editor.putBoolean(preference.getKey(), (Boolean) value);
        } else if (value instanceof String) {
            editor.putString(preference.getKey(), (String) value);
        } else {
            ByteArrayOutputStream outStream = new ByteArrayOutputStream();
            ObjectOutputStream out = new ObjectOutputStream(outStream);

            out.writeObject(value);
            out.flush();
            out.close();

            editor.putString(preference.getKey(), outStream.toString());
            outStream.flush();
            outStream.close();
        }
    }
}

File: preftils/src/main/java/com/hubble/preftils/PreferenceUtils.kt
package com.hubble.preftils

import android.content.SharedPreferences
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Wrapper class for handling SharedPreferences. Make sure your custom Kotlin classes have the @Serializable annotation, otherwise you will get runtime errors:
 *
 *     kotlinx.serialization.SerializationException: Serializer for class 'X' is not found.
 *     Please ensure that class is marked as '@Serializable' and that the serialization compiler plugin is applied.
 *
 * @property[key] the SharedPreference Key
 * @property[default] the default value of the preference, also provides concrete type-checkable type
 */
@Serializable
data class Preference<T: Any>(val key: String, val default: T)

/**
 * Retrieves the value of a preference in a type-safe way
 *
 * Propagates errors thrown by Json.decodeFromString() if decoding is impossible
 *
 * @param preference the Preference to retrieve
 *
 * @throws SerializationException in case of any decoding-specific error
 * @throws IllegalArgumentException if the decoded input is not a valid instance of [T]
 */
inline fun <reified T: Any> SharedPreferences.get(preference: Preference<T>, default: T = preference.default): T {
    return when (preference.default) {
        //Only Preferences type not supported is StringSet
        is String -> this.getString(preference.key, default as String) as T
        is Int -> this.getInt(preference.key, default as Int) as T
        is Long -> this.getLong(preference.key, default as Long) as T
        is Boolean -> this.getBoolean(preference.key, default as Boolean) as T
        is Float -> this.getFloat(preference.key, default as Float) as T
        else -> {
            if (preference.default::class.java.isAnnotationPresent(Serializable::class.java)) {
                val str = this.getString(preference.key, null) ?: return default
                return Json.decodeFromString<T>(str)
            }

            throw IllegalArgumentException("Unsupported type: ${preference.default::class.qualifiedName}")
        }
    }
}

/**
 * Sets the value of a preference in a type-safe way
 *
 * Propagates errors thrown by Json.encodeToString() if encoding of [value] is impossible
 *
 * @param[preference] The preference to save
 * @param[value] The value to set the preference to
 *
 * @throws SerializationException in case of any encoding-specific error
 * @throws IllegalArgumentException if the encoded input is not a valid instance of [T]
 */
inline fun <reified T: Any> SharedPreferences.Editor.put(preference: Preference<T>, value: T) {
    when (value) {
        //Only Preferences type not supported is StringSet
        is String -> this.putString(preference.key, value)
        is Int -> this.putInt(preference.key, value)
        is Long -> this.putLong(preference.key, value)
        is Boolean -> this.putBoolean(preference.key, value)
        is Float -> this.putFloat(preference.key, value)
        else -> {
            if (preference.default::class.java.isAnnotationPresent(Serializable::class.java)) {
                this.putString(preference.key, Json.encodeToString(value))
                return
            }
            throw IllegalArgumentException("Unsupported type: ${preference.default::class.qualifiedName}")
        }
    }
}
